kvbc/src/categorization/immutable_kv_category.cpp

```diff
@@ -0,0 +1,172 @@
+    auto &key = node.key();
+    auto &value = node.mapped();
+
+    // Calculate hashes (optionally) before moving them.
+    Hash key_hash;
+    Hash value_hash;
+    if (update.calculate_root_hash && !value.tags.empty()) {
+      key_hash = hash(key);
+      value_hash = hash(value.data);
+    }
+
+    // Persist the key-value.
+    batch.put(cf_, key, serialize(ImmutableValue{block_id, std::move(value.data)}));
+
+    // Move the key and the tags to the update info and (optionally) update hashes per tag.
+    auto &key_tags = update_info.tagged_keys.emplace(std::move(key), std::vector<std::string>{}).first->second;
+    for (auto &&tag : value.tags) {
+      if (update.calculate_root_hash) {
+        updateTagHash(tag, key_hash, value_hash, tag_hashers);
+      }
+      key_tags.push_back(std::move(tag));
+    }
+  }
+
+  // Finish hash calculation per tag.
+  if (update.calculate_root_hash) {
+    finishTagHashes(tag_hashers, update_info);
+  }
+  return update_info;
+}
+
+void ImmutableKeyValueCategory::deleteGenesisBlock(BlockId,
+                                                   const ImmutableUpdatesInfo &updates_info,
+                                                   storage::rocksdb::NativeWriteBatch &batch) {
+  deleteBlock(updates_info, batch);
+}
+
+void ImmutableKeyValueCategory::deleteLastReachableBlock(BlockId,
+                                                         const ImmutableUpdatesInfo &updates_info,
+                                                         storage::rocksdb::NativeWriteBatch &batch) {
+  deleteBlock(updates_info, batch);
+}
+
+void ImmutableKeyValueCategory::deleteBlock(const ImmutableUpdatesInfo &updates_info,
+                                            storage::rocksdb::NativeWriteBatch &batch) {
+  for (const auto &kv : updates_info.tagged_keys) {
+    batch.del(cf_, kv.first);
+  }
+}
+
+std::optional<ImmutableValue> ImmutableKeyValueCategory::get(const std::string &key) const {
+  const auto ser = db_->get(cf_, key);
+  if (!ser) {
+    return std::nullopt;
+  }
+  auto value = ImmutableValue{};
+  deserialize(*ser, value);
+  return value;
+}
+
+std::optional<KeyValueProof> ImmutableKeyValueCategory::getProof(const std::string &tag,
+                                                                 const std::string &key,
+                                                                 const ImmutableUpdatesInfo &updates_info) const {
+  auto it = updates_info.tagged_keys.find(key);
+  if (it == updates_info.tagged_keys.cend()) {
+    return std::nullopt;
+  }
+
+  auto value = get(key);
+  ConcordAssert(value.has_value());
+
+  auto proof = KeyValueProof{};
+  proof.block_id = value->block_id;
+  proof.key = key;
+  proof.value = std::move(value->data);
+
+  auto i = std::size_t{0};
+  for (const auto &[update_key, update_key_tags] : updates_info.tagged_keys) {
+    if (update_key == key) {
+      proof.key_value_index = i;
+      continue;
+    }
```

Comment:
Why the key is treated differently and not hashed and inserted to ```ordered_complement_kv_hashes``` as the other keys.
Eventually, it will be done in ```calculateRootHash`` if I understand correctly

Reply:
The intention is for the client to calculate the actual hash of the key-value being proven, instead of relying on a hash calculated here.